cmd/bedFilter/bedFilter.go

```diff
@@ -7,75 +7,79 @@ import (
 	"fmt"
 	"github.com/vertgenlab/gonomics/bed"
 	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/exception"
 	"github.com/vertgenlab/gonomics/fileio"
 	"github.com/vertgenlab/gonomics/numbers"
 	"log"
-	"strings"
+	"math"
+	"math/rand"
 )
 
-func bedFilter(infile string, outfile string, minScore int, maxScore int, minLength int, maxLength int, minStart int, maxStart int, minEnd int, maxEnd int, chrom string) {
-	var outlist []bed.Bed
-	var line string
-	var startNum, endNum, length int
-	var doneReading bool = false
+func bedFilter(s Settings) {
+	common.RngSeed(s.RandSeed, s.SetSeed)
+	var length int
 	var pass bool = false
-	var numFields int
-	var current bed.Bed
-	file := fileio.EasyOpen(infile)
-	defer file.Close()
+	var r float64
+	bedChan := bed.GoReadToChan(s.InFile)
+	out := fileio.EasyCreate(s.OutFile)
 
-	for line, doneReading = fileio.EasyNextRealLine(file); !doneReading; line, doneReading = fileio.EasyNextRealLine(file) {
+	for curr := range bedChan {
 		pass = true
-		words := strings.Split(line, "\t")
-		startNum = common.StringToInt(words[1])
-		endNum = common.StringToInt(words[2])
-		length = endNum - startNum
-		if len(words) > 4 {
-			if common.StringToInt(words[4]) < minScore {
+		length = curr.ChromEnd - curr.ChromStart
+		if curr.FieldsInitialized > 4 {
+			if curr.Score < s.MinScore {
 				pass = false
 			}
-			if common.StringToInt(words[4]) > maxScore {
+			if curr.Score > s.MaxScore {
 				pass = false
 			}
 		}
-		if length < minLength {
+		if length < s.MinLength {
 			pass = false
 		}
-		if length > maxLength {
+		if length > s.MaxLength {
 			pass = false
 		}
-		if startNum < minStart {
+		if curr.ChromStart < s.MinStart {
 			pass = false
 		}
-		if startNum > maxStart {
+		if curr.ChromStart > s.MaxStart {
 			pass = false
 		}
-		if endNum < minEnd {
+		if curr.ChromEnd < s.MinEnd {
 			pass = false
 		}
-		if endNum > maxEnd {
+		if curr.ChromEnd > s.MaxEnd {
 			pass = false
 		}
-		if chrom != "" {
-			if words[0] != chrom {
+		if s.MinNameFloat != -1*math.MaxFloat64 {
+			if common.StringToFloat64(curr.Name) < s.MinNameFloat {
 				pass = false
 			}
 		}
-		if pass {
-			if len(words) == 3 {
-				current = bed.Bed{Chrom: words[0], ChromStart: startNum, ChromEnd: endNum}
-				numFields = numbers.Max(3, numFields)
-			} else if len(words) == 4 {
-				current = bed.Bed{Chrom: words[0], ChromStart: startNum, ChromEnd: endNum, Name: words[3]}
-				numFields = numbers.Max(4, numFields)
-			} else {
-				current = bed.Bed{Chrom: words[0], ChromStart: startNum, ChromEnd: endNum, Name: words[3], Score: common.StringToInt(words[4])}
-				numFields = numbers.Max(5, numFields)
+		if s.MaxNameFloat != math.MaxFloat64 {
+			if common.StringToFloat64(curr.Name) > s.MaxNameFloat {
+				pass = false
+			}
+		}
+		if s.Chrom != "" {
+			if curr.Chrom != s.Chrom {
+				pass = false
+			}
+		}
+		if s.SubSet < 1.0 {
+			r = rand.Float64()
```

Comment:
This step can be expensive.  Maybe we don't need to do it if pass is already false?